www/js/lib/uiUtil.js

```diff
@@ -46,30 +46,49 @@ define(rqDef, function(settingsStore) {
      * @param {Function} callback An optional function to call with the URI
      */
     function feedNodeWithBlob(node, nodeAttribute, content, mimeType, callback) {
+        var insertBlob = function (blob) {
+            var url = URL.createObjectURL(blob);
+            if (callback) callback(url);
+            node.addEventListener('load', function () {
+                URL.revokeObjectURL(url);
+            });
+            node.setAttribute(nodeAttribute, url);
+        };
         // Decode WebP data if the browser does not support WebP and the mimeType is webp
         if (webpMachine && /image\/webp/i.test(mimeType)) {
             // DEV: Note that webpMachine is single threaded and will reject an image if it is busy
             // However, the loadImagesJQuery() function in app.js is sequential (it waits for a callback
             // before processing another image) so we do not need to queue WebP images here
-            webpMachine.decode(content).then(function (uri) {
-                // DEV: WebpMachine.decode() returns a data: URI
-                // We callback before the node is set so that we don't incur slow DOM rewrites before processing more images
-                if (callback) callback(uri);
-                node.setAttribute(nodeAttribute, uri);
+            var canvas = document.createElement('canvas');
+            webpMachine.decodeToCanvas(canvas, content).then(function () {
+                if (canvas.msToBlob) {
```

Comment:
Is it necessary to distinguish these 3 cases?
We might simply do `webpMachine.replaceImageWithCanvas(node, canvas);` in all cases? (if it's working well), and call the callback (if set) of course.
Do you see a side effect of replacing images by canvas on all browsers that do not support webp?
It would simplify the code, and might be a bit faster (from what I understood, it skips the conversion to/from base64)

Reply:
I kept it in there because it was a working method for IE11 and I was hoping to fine an analogous Firefox-specific one for IceCat.

The only reason to keep the IE11 method, and possibly the generic one, is that the one targeted at IceCat is very much a workaround: we have to guess at which attributes to copy from the `<img>` tag, and it's not very generic. For example, do we/can we copy alt text? What about a ZIM in the future that has image dimensions in `em` and not in points, etc.?

Of course it's only a tiny number of browsers that don't support WebP images, so maybe we're not sacrificing much if we use a single function. Having said that, the IE11 code is only a couple of lines. It would be worth seeing if there is an appreciable difference in IE11. Inserting a node into the DOM and creating a blob probably take the same amount of time. I also need to test on Edge Classic / UWP, as this impacts on the UWP app (on Windows Mobile, WebP isn't supported either). I suspect the `msToBlob` method works on Edge Classic.

I don't feel strongly about this, unless there is a significant performance loss.

Reply:
Plus there is the scrolling issue with inserted canvases: it certainly affects IceCat. I haven't tested on IE11. If the mouse is hovering over an image, it's not possible to scroll with mouse wheel or trackpad. It makes me think we should only use this method for cases that absolutely need it.

Reply:
OK, that makes sense.
The scrolling issue, and the risk to badly copy image attributes are enough to keep the original method for browsers that properly support it.

Problem is how to detect that.
On IceCat 60.7, there is no error. The anti-fingerprinting feature simply alters the canvas reading: that's hard to detect.
Maybe we should simply add a setting to use canvas or not (disabled by default).
This is probably a very rare case: we're talking about old browsers (that do not support webp) with a bad anti-fingerprinting feature. I say "bad" because the problem does not occur on latest Firefox with [CanvasBlocker](https://addons.mozilla.org/en-US/firefox/addon/canvasblocker/) extension (and with webp manually disabled)

Reply:
The nearest I could get to targeting it at IceCat was to use `if (canvas.mozGetAsFile)`. This is a deprecated function and any modern Firefox won't even go near this code because it supports WebP natively.

So this is likely to trap: only a browser that does not support WebP and is Mozilla-based (realistically the only browser type anyone would try to mod for privacy purposes, I think). It's not perfect, but it does apply to IceCat.

It might be worth checking if this accidentally traps Firefox OS, as presumably that's not so desirable. I can't remember if FFOS supports WebP natively. Maybe there's a way to get FFOS to run the native `canvas.toBlob()`, but I realize we're getting quite close to browser sniffing here, which we don't want to do.

Reply:
I'll test on Firefox OS when we'll have something working.
It's based on Gecko 44, so it does not support webp, see https://caniuse.com/webp
According to https://caniuse.com/mdn-api_htmlcanvaselement_mozgetasfile, it supports `mozGetAsFile`, so would probably be trapped by your test...
This test would trap any browser based on Gecko >=4 and <=65 (or <=73 and with webp manually disabled).
I think it's unfortunately too broad.

FFOS should support `canvas.toBlob()`, according to https://caniuse.com/mdn-api_htmlcanvaselement_toblob , and webp-hero already works well on it, with the current kiwix-js.

Reply:
There would probably be a way to properly detect that: take a small known webp image binary, make webp-hero convert it, and compare the result with the expected one.
If it's different, it means we have to fallback to keep the canvas.
This could be tested once, and displayed in the API settings

Reply:
Interesting, I like the idea. We currently do something similar already with WebHero to test whether the browser understands WebP (so we can load the polyfill conditionally, as it is quite heavy in terms of MB). Doing so in the same place would be easy.